proxy/rpc/np_gas_price.py

```diff
@@ -140,3 +171,198 @@ async def get_neon_gas_price(
+
+        # Let's clamp num_blocks to not create excessive load - Infura and Alchemy currently do the same.
+        block_cnt = min(block_cnt, _FEE_HISTORY_MAX_NUM_BLOCKS)
+
+        # Fetching the data about compute units on Solana.
+        cu_price_data_list: list[PriorityFeePercentiles] = list()
+        if is_reward_list:
+            cu_price_data_list = await self._db.get_historical_priority_fees(block_cnt, latest_slot_req)
+
+        # Determine the earliest_block in the response.
+        # If cu_price_data_list is requested and available, take as an earliest block slot.
+        # Otherwise, take `block_cnt` blocks backwards starting from what's requested.
+        # Also, let's clamp it against the earliest neon block so the response makes sense.
+        earliest_block_slot_resp: int = (
+            cu_price_data_list[-1].block_slot if cu_price_data_list else latest_slot_req - block_cnt + 1
+        )
+        earliest_neon_block: NeonBlockHdrModel = await self._db.get_earliest_block()
+        earliest_block_slot_resp = max(earliest_block_slot_resp, earliest_neon_block.slot)
+
+        # Fetching the data about base fees on Neon.
+        # The source of base_fee_per_gas data is either from the mempool (for the most recent blocks),
+        # or from the DB (for historical blocks).
+        # The reason for that is twofold:
+        # 1. Sparse non-empty Neon blocks. In case the User requested recent data, an operator returns recent
+        #   prices for the gas token regardless of what was set in Neon transactions.
+        # 2. Different incentives for different operators. Some operators may set the operator profit margin lower,
+        #   so the only way for them to benefit from that is if we return recent suggested token gas price from
+        #   their mempool and not from the transactions.
+        # However, the recent pricing (from the mempool) is timestamp-based while DB data is block_slot-based,
+        # so we need to do some heuristic conversion.
+        # The assumption we take here for pricing from the mempool: it corresponds to the most recent block_slot,
+        # even though it's not strictly accurate all the time (requests to Pyth may return errors sometimes).
+        base_fee_data_list: list[BlockFeeGasData] = list()
+        # First, take the token gas price data from the mempool (without querying the DB).
+        mempool_basefee_gas_prices: MpRecentGasPricesModel = await self._server.get_recent_gas_prices_list(ctx)
+        recent_basefee_price_list: list[MpGasPriceTimestamped] = mempool_basefee_gas_prices.token_gas_prices
+        if recent_basefee_price_list:
+            # Construct entries into base_fee_data_list from in-memory mempool-based recent token gas prices.
+            for gas_price in reversed(recent_basefee_price_list):
+                # Heuristically assign the block slot.
+                ephemeral_block_slot = mempool_basefee_gas_prices.get_ephemeral_block_slot(
```

Comment:
IMHO these calculations can be simplified.
Mempool has access to Solana.
So, you can get the latest slot and link it to the gas-price.
And here, you can just fill non-existent slots with the same gas-price.